packages/@glimmer/core/src/render-component/index.ts

```diff
@@ -0,0 +1,130 @@
+
+import { RootReference, PathReference } from '@glimmer/reference';
+import { ComponentFactory } from '../managers/component/custom';
+
+import { SimpleElement } from '@simple-dom/interface';
+
+export interface RenderComponentOptions {
+  element: Element;
+  args?: Dict<unknown>;
+  meta?: unknown;
+}
+
+type ResolveFn = () => void;
+type RejectFn = (error: Error) => void;
+
+let renderNotifiers: Array<[ResolveFn, RejectFn]> = [];
+
+export function didRender(): Promise<void> {
+  if (scheduled) {
+    return new Promise((resolve, reject) => {
+      renderNotifiers.push([resolve, reject]);
+    });
+  }
+  return Promise.resolve();
+}
+
+async function renderComponent(
+  ComponentClass: ComponentFactory,
+  options: RenderComponentOptions
+): Promise<void>;
+async function renderComponent(
+  ComponentClass: ComponentFactory,
+  element: HTMLElement
+): Promise<void>;
+async function renderComponent(
+  ComponentClass: ComponentFactory,
+  optionsOrElement: RenderComponentOptions | HTMLElement
+): Promise<void> {
+  const options: RenderComponentOptions =
+    optionsOrElement instanceof HTMLElement ? { element: optionsOrElement } : optionsOrElement;
+
+  const { element, args } = options;
+
+  const iterator = getTemplateIterator(ComponentClass, element, EnvironmentImpl.create(), args);
+  const result = iterator.sync();
+  results.push(result);
+}
+
+export default renderComponent;
+
+const results: RenderResult[] = [];
+
+setPropertyDidChange(scheduleRevalidation);
+
+let scheduled = false;
+function scheduleRevalidation(): void {
+  if (scheduled) {
+    return;
+  }
+
+  scheduled = true;
+  setTimeout(() => {
+    scheduled = false;
+    try {
+      revalidate();
+      renderNotifiers.forEach(([resolve]) => resolve());
+    } catch (err) {
+      renderNotifiers.forEach(([, reject]) => reject(err));
+    }
+
+    renderNotifiers = [];
+  }, 0);
+}
+
+function revalidate(): void {
+  for (const result of results) {
+    const { env } = result;
+    env.begin();
+    result.rerender();
+    env.commit();
+  }
+}
+
+const resolver = new RuntimeResolver();
+const context = JitContext(new CompileTimeResolver(resolver));
+
+function dictToReference(dict?: Dict<unknown>): Dict<PathReference> {
+  if (!dict) {
+    return {};
+  }
+
+  return Object.keys(dict).reduce((acc, key) => {
+    acc[key] = new RootReference(dict[key]);
+    return acc;
+  }, {} as Dict<PathReference>);
+}
+
+export function getTemplateIterator(
+  ComponentClass: ComponentFactory,
+  element: Element | SimpleElement,
+  env: Environment,
+  componentArgs?: Dict<unknown>
+): TemplateIterator {
+  const runtime = CustomJitRuntime(resolver, context, env);
+  const builder = clientBuilder(runtime.env, {
+    element,
+    nextSibling: null,
+  } as GlimmerCursor);
+
+  const handle = resolver.registerRoot(ComponentClass);
+
+  return renderJitComponent(runtime, builder, context, 0, handle, dictToReference(componentArgs));
```

Comment:
How would this evolve if we want to support AoT mode with `renderComponent`? One thing that <= 1.0.0 did was forced the application to decide how the VM worked e.g. AOT vs JIT, Sync Render vs Async Renderer. It seems like we would have to cut 3.0.0 to add that feature.

Reply:
We should discuss this more for sure before we cut the final v2 release. I know @wycats and I were talking about trying to unify the AOT and JIT models more, and we’ll have to do that to support lazy bundling and such. 

I wonder if it could be more of a compiler decision, and if we could expand the `renderComponent` API to support additional metadata for AOT compilation without needing to make breaking changes.

Reply:
Yea I would assume that having first class values makes a lot of the resolver indirection.

I think the thing I think we should do here is to require `renderComponent` to take the mode specific bits. Its not a big deal for the server-side have defaults but I think we should continue to have code for configuration to ensure that rendering can happen different ways, but not pay the tax of downloading the code for the unused modes.

For example we could do something like capabilities:

```js
renderComponent(Foo, {
  element: document.getElementById('foo'),
  mode: JIT()
});
```

Then we would have mode mappings like the following:

```js
const CURRENT_MODE = new WeakMap();

const JITSetup = {
    renderer: renderJitComponent,
    runtime: CustomJitRuntime,
     // ...
}

const AoTSetup = {
    renderer: renderAoTComponent,
    runtime:   AotRuntime,
     // ...
}

const JITMode =  {
  mode: 'JIT',
};

// All the JIT bits
function JIT() {
  CURRENT_MODE.set(JITMode, JITSetup);
  return JITMode;
}

const AoTMode =  {
  mode: 'AoT',
};

// All the JIT bits
function AoT() {
  CURRENT_MODE.set(AoTMode, AoTSetup);
  return AoTMode;
}

renderComponent(Component, options) {
  const { mode } = options.mode;
  const { Renderer, Runtime } = CURRENT_MODE.get(options.mode);
  
  if (mode === 'JIT) {
    // JIT specific setup
  } else {
    // AoT specific setup 
  }
}

```